src/Features/LanguageServer/Protocol/Handler/BufferedProgress.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Helper type to allow command handlers to report data either in a streaming fashion (if a client supports that),
+    /// or as an array of results.  This type is thread-safe in the same manner that <see cref="IProgress{T}"/> is
+    /// expected to be.  Namely, multiple client can be calling <see cref="IProgress{T}.Report(T)"/> on it at the same
+    /// time.  This is safe, though the order that the items are reported in when called concurrently is not specified.
+    /// </summary>
+    internal struct BufferedProgress<T> : IProgress<T[]>, IDisposable
```

Comment:
Potentially this could also implement `IProgress<T>` to provide the best of all worlds?

Reply:
Turns out this was required for FAR to work in tests, so I've added it in my PR.